### PR TITLE
Add mate pair phase change detection

### PR DESCRIPTION
**Pull request by JimmyLiu8351**

### CODE REVIEW

File: readcomb/phase_change_filter.py [hunk 1]
```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
...
+            index += cigar_tuple[1]\
+
+        # 2 is an insertion, add a gap to the query
+        elif cigar_tuple[0] == 2:
+            segment += '-' * cigar_tuple[1]
+            index += cigar_tuple[1]
+
         else:
-            print('oops forgot to consider this: ' + str(cigar_tuple))
+            print('forgot to consider this: ' + str(cigar_tuple))
```
Comment (aays) [inline readcomb/phase_change_filter.py:120]: TODO: reformat as exception

File: readcomb/phase_change_filter.py [hunk 2]
```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
...
+        each snp is a cyvcf2 variant object
+    segment: string sequence that is the segment built by the function 
+    '''
+    
     snp_lst = []
     
+    cigar_tuples = record.cigartuples
+    
+    # check for no alignment
+    if cigar_tuples == None:
```
Comment (aays) [inline readcomb/phase_change_filter.py:139]: TODO: find read raising this issue

File: readcomb/phase_change_filter.py [hunk 3]
```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
...
         # hard clipping so we don't skip it and we don't add anything
         elif cigar_tuple[0] == 5:
             continue
-        
-        # if it is a match (0) or an insertion, then just add it onto the segment 
-        elif cigar_tuple[0] in [0, 1]:
-            segment += query_sequence[index:index+cigar_tuple[1]]
+
+        # 0 is a match, just add it onto the segment 
+        elif cigar_tuple[0] == 0:
```
Comment (JimmyLiu8351) [inline readcomb/phase_change_filter.py:105]: Switch to list concatenation to be better with Pypy

File: readcomb/phase_change_filter.py [hunk 4]
```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
...
-                'no_snp_reads']
-                out_values = [phase_change_reads, total_reads, no_match_reads,
-                        no_match_phase_change_reads, no_snp_reads]
-                f.write(','.join(fieldnames) + '\n')
-                f.write(','.join([str(n) for n in out_values]) + '\n')
+    '''.format(phase_change_counter, all_seq_counter, phase_change_counter / all_seq_counter, 
+        phase_change_mate_pair_counter,
+        no_match_counter, all_seq_counter - seq_with_snps_counter,
+        (all_seq_counter - seq_with_snps_counter) / all_seq_counter)
+    )
```
Comment (JimmyLiu8351) [inline readcomb/phase_change_filter.py:299]: Bring back log code

**Comment by aays** [comment]
One more thing - I'd very much prefer we use `tqdm` around iterables where possible. It prints out progress bars as well as iterations/second metrics and I think it's pretty key here. The import is `from tqdm import tqdm` and we wrap iterables with `tqdm` to have it print the bars (e.g. `for query_name in tqdm(pairs)`). This is also just generally helpful for rich output while the script runs.

File: readcomb/phase_change_filter.py [hunk 5]
```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
...
+        snps = check_snps(vcf, record.reference_name, 
+                        record.reference_start,
+                        record.reference_start + record.query_alignment_length)
+        
+        snp_lst = phase_detection(snps, segment, record)
+        
+        # initialize second record if there is a matepair
+        if pairs[query_name][1]:
+            
+            seq_with_snps_counter += 1
```
Comment (aays) [inline readcomb/phase_change_filter.py:236]: Is this counter specifically to count the number of reads with pairs? Bc if so it should have a different name

File: readcomb/phase_change_filter.py [hunk 6]
```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
...
+    no_match_counter = 0
+    phase_change_counter = 0
+    phase_change_mate_pair_counter = 0
+    seq_with_snps_counter = 0
+    all_seq_counter = 0
+    
+    # get mate pairs
+    pairs = cache_pairs(bam_file_obj)
+    
+    for query_name in pairs:
```
Comment (aays) [inline readcomb/phase_change_filter.py:217]: Here's what I was thinking for this part:

```python
for query_name in pairs:
        
        all_seq_counter += 1
        record_counter = 0
        snp_lst = []
        
        for record in pairs[query_name]:
            if record:
                record_counter += 1
            
                # analyze cigar string
                segment = cigar(record)
                            
                snps = check_snps(vcf, record.reference_name, 
                                record.reference_start,
                                record.reference_start + record.query_alignment_length)
                
                snp_lst += phase_detection(snps, segment, record)

                if 'N' in snp_lst:
                    no_match_counter += 1

                    if 'no_match' in mode:                        
                        f_obj.write(record)

        #phase_change_counter update
        if '1' in snp_lst and '2' in snp_lst:
            phase_change_counter += 1

            if 'phase_change' in mode:                        
                if record_counter == 1:
                    f_obj.write(pairs[query_name][0])
                elif record_counter == 2:
                    for record in pairs[query_name]:
                        f_obj.write(record)

```

It makes the code a lot simpler by consolidating things into a single snp list, and then just doing a few checks for whether the second record exists where needed. We can chat more about it on Thursday

File: readcomb/phase_change_filter.py (hunk 1 above)
Comment (JimmyLiu8351) [reply readcomb/phase_change_filter.py:120]: Implemented as exception
```python
raise Exception('No condition for tuple ' + str(cigar_tuples.index(cigar_tuple)) + ' of ' + str(cigar_tuples))
```

File: readcomb/phase_change_filter.py (hunk 2 above)
Comment (JimmyLiu8351) [reply readcomb/phase_change_filter.py:139]: Created text file called unaligned_reads.txt with around 360 unaligned reads that you could look at, will be included in the next commit

File: readcomb/phase_change_filter.py (hunk 3 above)
Comment (JimmyLiu8351) [reply readcomb/phase_change_filter.py:105]: Implemented

File: readcomb/phase_change_filter.py (hunk 4 above)
Comment (JimmyLiu8351) [reply readcomb/phase_change_filter.py:299]: Added with new counter variable names

```python
    if log:
        needs_header = True
        if os.path.isfile(log):	
            needs_header = False	
        with open(log, 'a') as f:	
            if needs_header:	
                fieldnames = ['phase_change_reads', 'total_reads',	
                'no_match_reads', 'phase_change_across_mate_pairs',	
                'no_snp_reads']	
                out_values = [phase_change_counter, all_seq_counter, no_match_counter,	
                        phase_change_mate_pair_counter, all_seq_counter - seq_with_snps_counter]	
                f.write(','.join(fieldnames) + '\n')	
                f.write(','.join([str(n) for n in out_values]) + '\n')
```

File: readcomb/phase_change_filter.py (hunk 5 above)
Comment (JimmyLiu8351) [reply readcomb/phase_change_filter.py:236]: Oh yeah, that's a relic left over from before I overhauled matepair_recomb, will update this

File: readcomb/phase_change_filter.py (hunk 6 above)
Comment (JimmyLiu8351) [reply readcomb/phase_change_filter.py:217]: This is actually really clean, I guess I really tunnel visioned when I rewrote this. I'll change the code and test it

**Pull request merged**